### ISSUE TITLE
Fail to upload to MEGA

### ISSUE THREAD

**Issue by joseph1125** [qgustavor/mega#44]
When I try to upload files to MEGA, the error is indicated as below

CODE:
```
private uploadFile(filePath: string, parentFolder: MutableFile): Promise<Boolean> {
    const fileName = path.basename(filePath);
    return new Promise((resolve, reject) => {
      const readStream = fs.createReadStream(filePath);
      readStream.on('error', error => reject(error));
      readStream.on('end', () => {
        resolve(true);
      });
      readStream.pipe(parentFolder.upload({
        name: fileName,
        size: fs.statSync(filePath).size
      }));
    })
  }
```

ERROR: 
c:\Users\zx261\Documents\projects\project\node_modules\megajs\dist\main.node-cjs.js:384

  key.copy(newKey);

      ^

TypeError: Cannot read property 'copy' of undefined

    at unmergeKeyMac (c:\Users\zx261\Documents\projects\project\node_modules\megajs\dist\main.node-cjs.js:384:7)

    at getCipher (c:\Users\zx261\Documents\projects\project\node_modules\megajs\dist\main.node-cjs.js:308:18)

    at checkCallbacks (c:\Users\zx261\Documents\projects\project\node_modules\megajs\dist\main.node-cjs.js:1582:7)

    at streamToCb (c:\Users\zx261\Documents\projects\project\node_modules\megajs\dist\main.node-cjs.js:1789:11)

    at Request.<anonymous> (c:\Users\zx261\Documents\projects\project\node_modules\megajs\dist\main.node-cjs.js:26:7)

    at Request.emit (events.js:187:15)

    at IncomingMessage.<anonymous> (c:\Users\zx261\Documents\projects\project\node_modules\request\request.js:1083:12)

    at Object.onceWrapper (events.js:273:13)

    at IncomingMessage.emit (events.js:187:15)

    at endReadableNT (_stream_readable.js:1094:12)

**Comment by qgustavor** [comment]
I just uploaded [this file](https://mega.nz/#F!t81mUQ7Q!FrouxoInsumoAGRADECIDA) using this code:

```js
folder.upload({ name, size }, fs.createReadStream(filePath), function (err, file) {
  if (err) throw err
  console.log('File was uploaded')
})
```

It comes from my testing script: takes a random image from my wallpapers folder and upload to a folder with a random generated name and share it.

I remember I got this error while testing: for some reason the final key (which is, the random generated key and the message authentication code) isn't defined on the [checkCallbacks](https://github.com/qgustavor/mega/blob/57a4615199f0d0e80b0fa69e3256ce0474a63218/lib/mutable-file.js#L143) function.

As I said above it works for me. Maybe there's a race condition between the encrypter generating the MAC and the checkCallbacks call. I will put a "help wanted" label and test again piping to the encrypter stream instead of using as the `body` argument.

**Comment by joseph1125** [comment]
tried your code but still not working, stucked at the same error again.

**Comment by joseph1125** [comment]
I also tried to use Buffer, the error doesn't show up anymore, however, it is stucked without spilling error message and no file is uploaded.

```
return new Promise((resolve, reject) => {
      parentFolder.upload(
        {
          name: fileName,
          size: fs.statSync(filePath).size
        },
        fs.readFileSync(filePath),
        (error, file) => {
          if (error) {
            reject(error);
          }
          resolve(true);
        }
      );
    });
```

**Comment by qgustavor** [comment]
What is the node version you're using?

**Comment by joseph1125** [comment]
V10

**Comment by qgustavor** [comment]
I was using v8.11.1. I switched to v10.12.0 and tested again some times:

1. The test script created [this folder](https://mega.nz/#F!0xEmzSAK!DitameOxenteCALVINISTA) then stopped working: for some reason the upload callback wasn't called;
2. The test script created [this folder](https://mega.nz/#F!Ylc2AShJ!AriscaEficazMORFOLOGIA) then the upload finished successfully;
3. The test script created [this folder](https://mega.nz/#F!BtVwXK7T!MunidoAmuadoFLUXOGRAMA) then stopped working again;
4. (now with a sniffer enabled) The test script created [this folder](https://mega.nz/#F!w9dAGQwZ!ChacalGaludoTECNOCRATA) and the upload finished successfully;
5. The test script created [this folder](https://mega.nz/#F!Q4dCGIRD!SurtarEsteioCONJECTURA) then stopped working again. The sniffer shows that four connections were used and the server don't returned the upload "hash";
6. The test script created [this folder](https://mega.nz/#F!cx9SmIBQ!AlentoMulataDESVENTURA) and the upload finished successfully;

I did the following changed on "mutable-file.js":

* De-duplicated chunk verification code;
* Added options to control how chunks work and the maximum number of connections;
* Added a `process.nextTick` so in case some node version calls `checkCallbacks` before `megaEncrypt`.

Then I tested again:

1. [It worked](https://mega.nz/#F!Q88yASAT!PrezamDecanoMONOPLEGIA);
2. [It worked](https://mega.nz/#F!l180QCTJ!PuerilCrassoBOQUETEIRA);
3. [It worked](https://mega.nz/#F!V41QEIwC!AlheioAdeptaCONCUNHADA);
4. [It worked](https://mega.nz/#F!o102XCQZ!EmendaPreulaTISIOLOGIA).

I will publish a new version with those changes, update the master branch and document those new options.

**Comment by qgustavor** [comment]
I just published v0.15.2: does it solve this issue?

**Comment by qgustavor** [comment]
Fixing #32 could make it easier: as uploading requires an account it's a bit hard to test it, but if you know some way to test uploading fixing this will get easier as we can test in multiple platforms and Node versions avoiding this kind of issue.

I'm using Node v8.12.0 and v10.12.0 on Windows 7. I just tested it in other platform (I think it's a unix system) and I got "TypeError: Cannot read property 'copy' of undefined". I don't think it might be a cross platform bug, so maybe the `node_modules` folder of the test script is corrupted: I created I new test script in other folder and tried again, then got the same bug.

Now that I can reproduce this bug I can debug it: ["end" function](https://github.com/qgustavor/mega/blob/ec83a33729f6c93837184c8879dfe4c58a5f473f/lib/crypto/index.js#L62-L66) isn't being called. "end" should be called by through. In two test folders the through library points to the same version. Comparing packages from those two folders most changes were minor changes. stream-skip@1.0.2 was updated to stream-skip@1.0.3 ([just to update documentation](https://github.com/Ivshti/stream-skip/commit/7c063ac709aaff1b8b92ee25aba6aab9d2a83f7c)), combined-stream@1.0.5 to combined-stream@1.0.7 (but it's not used when uploading).

I will keep testing it here...

**Comment by qgustavor** [comment]
As I said before my tests were working.  Can you reduce your problem in a
reproducible way? You can publish your code in a gist then tell what to do
to get those errors.

There's a issue about testing upload code. As uploading requires an account
it's a bit hard to test it. If you know some way to test uploading fixing
this will get easier as we can test in many platforms and Node versions.

**Comment by qgustavor** [comment]
v0.15.4 was released.

**Comment by joseph1125** [comment]
With the new update, errors pop out and the errors depend on the file

For  [z.zip](https://github.com/qgustavor/mega/files/2506186/z.zip), the error is

```
Error: Specified data size does not match: 6212497 !== 2293760 <=the later number always change
warning.js:18
    at Stream.source.on (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:1732:33)
    at Stream.emit (events.js:187:15)
    at Stream.handleEnd (d:\Projects\novels-manager\node_modules\duplexer\index.js:81:21)
    at Stream.emit (events.js:187:15)
    at Stream.end (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:347:10)
    at _end (d:\Projects\novels-manager\node_modules\through\index.js:65:9)
    at Stream.stream.end (d:\Projects\novels-manager\node_modules\through\index.js:74:5)
    at Stream.onend (internal/streams/legacy.js:42:10)
    at Stream.emit (events.js:187:15)
    at Stream.<anonymous> (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:54:10)
```

For [b.zip](https://github.com/qgustavor/mega/files/2506196/b.zip), the file did uploaded to MEGA, however, the following error pops out

```
Error: Attributes could not be decrypted with provided key.
main.node-cjs.js:1415
    at Function.unpackAttributes (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:1415:13)
    at MutableFile.decryptAttributes (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:1093:15)
    at MutableFile.loadMetadata (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:1085:12)
    at new MutableFile (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:1452:12)
    at Storage._importFile (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:2302:38)
    at api.request (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:1615:35)
    at Request.requestModule [as _callback] (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:983:7)
    at Request.self.callback (d:\Projects\novels-manager\node_modules\request\request.js:185:22)
    at Request.emit (events.js:182:13)
    at Request.<anonymous> (d:\Projects\novels-manager\node_modules\request\request.js:1161:10)
```

Thanks for your quick response.

**Comment by qgustavor** [comment]
The bug is on the [upload chunking code](https://github.com/qgustavor/mega/blob/5c866aad4468349f8765fe38a40e41d7ac182eb6/lib/mutable-file.js#L308). Can you check it?

**Comment by joseph1125** [comment]
> The bug is on the [upload chunking code](https://github.com/qgustavor/mega/blob/5c866aad4468349f8765fe38a40e41d7ac182eb6/lib/mutable-file.js#L308). Can you check it?

For z.zip and b.zip , the initialChunkSize is 131072

**Comment by qgustavor** [comment]
I linked just the first line, but chunking code goes from 308 to L381. 131072 is 128 * 1024 (128 KiB). You can read more about it here: https://mega.nz/doc (it's on "12.10 File encryption").

**Comment by qgustavor** [comment]
I wrote a partial reimplementation of MEGA servers. By now it handles login, file listing, uploading and sharing. I want to implement shared file download and shared folder listing.

While developing it I found some bugs:

* It's not handling errors: it don't stops uploading when the server returns an error;
* There's some error in the chunking code: it uploaded 1280 KiB from a file then stopped the upload midway;
* Folder sharing code is also not handling errors.

**Comment by qgustavor** [comment]
I just published v0.15.5.

**Comment by joseph1125** [comment]
Great news, now all those test files can be uploaded, however, after the successful upload, the following error is thrown

```
Error: Attributes could not be decrypted with provided key.
    at Function.unpackAttributes (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:1415:13)
    at MutableFile.decryptAttributes (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:1093:15)
    at MutableFile.loadMetadata (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:1085:12)
    at new MutableFile (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:1452:12)
    at Storage._importFile (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:2311:38)
    at api.request (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:1622:35)
    at Request.requestModule [as _callback] (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:983:7)
    at Request.self.callback (d:\Projects\novels-manager\node_modules\request\request.js:185:22)
    at Request.emit (events.js:182:13)
    at Request.<anonymous> (d:\Projects\novels-manager\node_modules\request\request.js:1161:10)
Error: Attributes could not be decrypted with provided key.
    at Function.unpackAttributes (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:1415:13)
    at MutableFile.decryptAttributes (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:1093:15)
    at MutableFile.loadMetadata (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:1085:12)
    at new MutableFile (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:1452:12)
    at Storage._importFile (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:2311:38)
    at api.request (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:1622:35)
    at Request.requestModule [as _callback] (d:\Projects\novels-manager\node_modules\megajs\dist\main.node-cjs.js:983:7)
    at Request.self.callback (d:\Projects\novels-manager\node_modules\request\request.js:185:22)
    at Request.emit (events.js:182:13)
    at Request.<anonymous> (d:\Projects\novels-manager\node_modules\request\request.js:1161:10)
```

**Comment by qgustavor** [comment]
> now all those test files can be uploaded

For some reason I don't remember anymore I wrote a batch-javascript polyglot: I will not reuse this messy code, instead I will polish the mock server, put it separate repository and publish it on NPM, then rewrite [the current AVA/nock](https://github.com/qgustavor/mega/blob/99de3825feaae36d94d82226b5bc1f68a4db6969/test/storage.test.js) tests using it.

I don't know what's a good way to config the mock server: it needs a folder where files will be uploaded, should who's using the server provide a folder or should it create a folder using `fs.mkdtemp`?

> the following error is thrown

Do you uploaded it to a shared folder? If so I still need to fix the code which handles files on shared folders. It's a bit quirky as when you share a folder the files on it get encrypted with the share key, not your account key. As tonistiigi's mega don't handled it that's what I understood when I implemented shared folders on issue #1.

**Comment by joseph1125** [comment]
> > now all those test files can be uploaded
> 
> For some reason I don't remember anymore I wrote a batch-javascript polyglot: I will not reuse this messy code, instead I will polish the mock server, put it separate repository and publish it on NPM, then rewrite [the current AVA/nock](https://github.com/qgustavor/mega/blob/99de3825feaae36d94d82226b5bc1f68a4db6969/test/storage.test.js) tests using it.
> 
> I don't know what's a good way to config the mock server: it needs a folder where files will be uploaded, should who's using the server provide a folder or should it create a folder using `fs.mkdtemp`?
> 
> > the following error is thrown
> 
> Do you uploaded it to a shared folder? If so I still need to fix the code which handles files on shared folders. It's a bit quirky as when you share a folder the files on it get encrypted with the share key, not your account key. As tonistiigi's mega don't handled it that's what I understood when I implemented shared folders on issue #1.

Yes, you are correct, I did uploaded them to a shared folder.

**Comment by qgustavor** [comment]
IIRC it only happens on upload, if you login again it will use the right key and decrypt it. I will close this issue and create other: #45. I don't plan to fix it too quick as I not pretend using this feature soon (unlike uploading).

**Comment by joseph1125** [comment]
Thanks for your help, will keep track of issue #45

**Comment by qgustavor** [comment]
As I said before I'm trying to finish the mock server (by the way someone knows english well can tell me if this name is ok?). Upload code broke (I got [this glitched image](https://i.imgur.com/L0Ei6yN.png) when uploading [this image](https://i.imgur.com/lnSyIYm.jpg)).

It failed on Node v10.13.0. When testing on v8.12.0 and v10.12.0 it uploaded without any issue. I will check Node changelog, check what changed and fix the upload code.

**Comment by joseph1125** [comment]
This name is good enough, very descriptive.

Gustavo Rodrigues <notifications@github.com> 於 2018年11月13日 週二 20:16寫道：

**Comment by qgustavor** [comment]
I found some edge cases and optimized the code (now it uses just one data handler). In the other hand when I tested using the real MEGA server it worked, so the corruption bug is on the mock server. I will commit those changes later.

**Comment by qgustavor** [comment]
No, it was a bug in the code that handled data that remained between one chunk and other. Seems old Node read files aligned to the chunk boundaries, so this error didn't happened on these. Fixed in 0.15.6 (aba6739).

*Now I can continue testing (and fixing) the mock server...*
